optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java

```diff
@@ -0,0 +1,318 @@
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import javax.xml.bind.Unmarshaller;
+import javax.xml.bind.UnmarshallerHandler;
+import javax.xml.bind.ValidationEvent;
+import javax.xml.bind.util.ValidationEventCollector;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Source;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMResult;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.transform.stream.StreamSource;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+
+import org.optaplanner.core.impl.io.OptaPlannerXmlSerializationException;
+import org.xml.sax.Attributes;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.XMLFilter;
+import org.xml.sax.XMLReader;
+import org.xml.sax.helpers.XMLFilterImpl;
+
+public final class GenericJaxbIO<T> implements JaxbIO<T> {
+    private static final int DEFAULT_INDENTATION = 2;
+
+    private static final String ERR_MSG_WRITE = "Unable to write the %s to XML.";
+    private static final String ERR_MSG_READ = "Unable to read the (%s) from XML.";
+
+    private final JAXBContext jaxbContext;
+    private final Marshaller marshaller;
+    private final Class<T> rootClass;
+    private final int indentation;
+
+    public GenericJaxbIO(Class<T> rootClass) {
+        this(rootClass, DEFAULT_INDENTATION);
+    }
+
+    public GenericJaxbIO(Class<T> rootClass, int indentation) {
+        Objects.requireNonNull(rootClass);
+        this.rootClass = rootClass;
+        this.indentation = indentation;
+        try {
+            jaxbContext = JAXBContext.newInstance(rootClass);
+            marshaller = jaxbContext.createMarshaller();
+            marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
+            marshaller.setProperty(Marshaller.JAXB_ENCODING, StandardCharsets.UTF_8.toString());
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format("Unable to create JAXB Marshaller for a root element class (%s).",
+                    rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+    }
+
+    @Override
+    public T read(Reader reader) {
+        Objects.requireNonNull(reader);
+        try {
+            return (T) createUnmarshaller().unmarshal(reader);
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format(ERR_MSG_READ, rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+    }
+
+    public T readAndValidate(Reader reader, String schemaResource) {
```

Comment:
As we agreed, we don't validate the input solver config XML. We use the validation just in tests to make sure the solver config XML is valid by the generated XSD.